packages/compass-aggregations/src/modules/aggregation.ts

```diff
@@ -204,9 +225,89 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
   return state;
 };
 
+const confirmWriteOperationIfNeeded = async (
+  instance: MongoDBInstance,
+  dataService: DataService,
+  namespace: string,
+  pipeline: Document[]
+) => {
+  if (!pipeline.length) {
+    return true;
+  }
+
+  const lastStageOperator = getStageOperator(pipeline[pipeline.length - 1]);
+  let typeOfWrite;
+
+  if (!lastStageOperator || !isOutputStage(lastStageOperator)) {
+    return true;
+  }
+
+  const destinationNamespace = getDestinationNamespaceFromStage(namespace, {
+    [lastStageOperator]: pipeline[pipeline.length - 1][lastStageOperator],
+  });
+
+  if (lastStageOperator === '$out') {
+    let isOverwritingCollection;
+
+    if (destinationNamespace) {
+      const { database, collection } = toNS(destinationNamespace);
+      isOverwritingCollection = !!(await instance.getNamespace({
+        dataService,
+        database,
+        collection,
+      }));
+    } else {
+      isOverwritingCollection = true;
+    }
+
+    typeOfWrite = isOverwritingCollection
+      ? WriteOperation.Overwrite
+      : WriteOperation.Create;
+  } else {
+    typeOfWrite = WriteOperation.Alter;
+  }
+
+  return await showConfirmation({
+    variant:
+      typeOfWrite === WriteOperation.Overwrite
+        ? ConfirmationModalVariant.Danger
+        : ConfirmationModalVariant.Default,
+    title: 'A write operation will occur',
+    description: runPipelineConfirmationDescription({
+      typeOfWrite,
+      stage: {
+        name: lastStageOperator,
+        link: WRITE_STAGE_LINK[
+          lastStageOperator as keyof typeof WRITE_STAGE_LINK
+        ],
+      },
+      ns: destinationNamespace,
+    }),
+    buttonText: 'Yes, run pipeline',
+    'data-testid': `write-operation-confirmation-modal`,
+  });
+};
+
 export const runAggregation = (): PipelineBuilderThunkAction<Promise<void>> => {
-  return (dispatch, getState, { pipelineBuilder, logger: { track } }) => {
-    const pipeline = getPipelineFromBuilderState(getState(), pipelineBuilder);
+  return async (
+    dispatch,
+    getState,
+    { pipelineBuilder, logger: { track }, instance, dataService }
+  ) => {
+    const state = getState();
+    const pipeline = getPipelineFromBuilderState(state, pipelineBuilder);
+
+    if (
+      !(await confirmWriteOperationIfNeeded(
+        instance,
+        dataService,
+        state.namespace,
+        pipeline
+      ))
+    ) {
+      return;
+    }
```

Comment:
It's an antipattern to store state in a variable inside thunk actions. Even though both calls here are sync, let's keep the pattern just so it's not confusing that we do one thing here, but a different one elsewhere in this code:

```suggestion
    const pipeline = getPipelineFromBuilderState(getState(), pipelineBuilder);

    if (
      !(await confirmWriteOperationIfNeeded(
        instance,
        dataService,
        getState().namespace,
        pipeline
      ))
    ) {
      return;
    }
```